### PR TITLE
47_regipol rewrite - several changes and new features (70+ commits). Please read the PR details.

### PR DESCRIPTION
**Pull request by Renato-Rodrigues**
General description:

**Main new features:** 
* Performance, consistency and code structure improvements driven by current and future demands from ECEMF and ARIADNE projects
* Improved regional emission tax target algorithm.
* Added support for PE, SE, FE and/or CCS targets achievable by policy taxes and subsidies. It can be interpreted as the necessary markups/markdowns to achieve some some tech or energy penetration target from the model perspective.
* Added support for PE or FE exogenously defined prices. It can be used to test the model behavior if a certain energy carrier has an specific price. The algorithm calculates and update endogenously an "implicit" tax or subsidy so the model always sees the desired exogenous price at its optimization decision.
* Added support for defining minimum VRE shares and maximum CCS values at REMIND native region level for ECEMF diagnostic scenarios.
*Added detailed FE energy carrier marginal prices and new FE balance after tax equation.

**Switches added and/or changed:** 

switch | change | related to | observation
-- | -- | -- | --
cm_emiMktTarget | added |   | new switch able to handle economy wide targets and emission specific   targets altogether
cm_implicitQttyTarget | added |   | new switch to force the model either through a tax or subsidy to reach an   specific demand target at PE, SE, FE or CCS level
cm_implicitPriceTarget | added |   | new switch to force the model to run with an exogenously defined FE   price, using an dynamically calculated tax or subsidy
cm_implicitPePriceTarget | added |   | new switch to force the model to run with an exogenously defined PE   price, using an dynamically calculated tax or subsidy
cm_VREminShare | added |   | minimum VRE share per country
cm_CCSmaxBound | added |   | maximum CCS per country
cm_quantity_regiCO2target | added |   | emission forced quantity for debugging purposes (forced by a hard bound   quantity equation instead of a tax based algorithm such as in   cm_emiMktTarget)
cm_postTargetIncrease | default value changed to 0 (from 2) | cm_emiMktTarget |  
cm_emiMktTargetDelay | added | cm_emiMktTarget | option for delayed price trajectory effects for latter targets
cm_prioRescaleFactor | added | cm_emiMktTarget | factor applied   to carbon tax rescale factor to prioritize short term targets in the initial   15 iterations
cm_loadFromGDX_implicitQttyTargetTax | added | cm_implicitQttyTarget | load tax information from gdx for policy runs
cm_regiCO2target | deprecated | cm_emiMktTarget | replaced by cm_emiMktTarget
cm_emiMktETS | deprecated | cm_emiMktTarget | replaced by cm_emiMktTarget
cm_emiMktETS_type | deprecated | cm_emiMktTarget |  
cm_ETS_postTargetIncrease | deprecated | cm_emiMktTarget |  
cm_ETS_post2055Increase | deprecated | cm_emiMktTarget |  
cm_emiMktES | deprecated | cm_emiMktTarget | replaced by cm_emiMktTarget
cm_emiMktES_type | deprecated | cm_emiMktTarget |  
cm_ESD_postTargetIncrease | deprecated | cm_emiMktTarget |  
cm_ESD_post2055Increase | deprecated | cm_emiMktTarget |  
cm_emiMktEScoop | deprecated | cm_emiMktTarget |  
cm_emiMktES2020price | deprecated | cm_emiMktTarget |  
cm_emiMktES2050 | deprecated | cm_emiMktTarget |  
cm_implicitFE | deprecated | cm_implicitQttyTarget | replaced by more general cm_implicitQttyTarget formulation
cm_implFETarget | deprecated | cm_implicitQttyTarget |  
cm_implFEExoTax | deprecated | cm_implicitQttyTarget |  
cm_inco0RegiFactor | added |   |  
cm_INNOPATHS_eni | renamed to cm_eni |   |  
cm_INNOPATHS_enb | renamed to cm_enb |   |  
cm_INNOPATHS_incolearn | renamed to cm_incolearn |   |  
cm_INNOPATHS_storageFactor | renamed to cm_storageFactor |   |  
cm_INNOPATHS_adj_seed | renamed to   cm_adj_seed |   |  
cm_INNOPATHS_adj_seed_cont | renamed to   cm_adj_seed_cont |   |  
cm_INNOPATHS_adj_coeff | renamed to   cm_adj_coeff |   |  
cm_INNOPATHS_adj_coeff_cont | renamed to   cm_adj_coeff_cont |   |  
cm_INNOPATHS_adj_seed_multiplier | renamed to   cm_adj_seed_multiplier |   |  
cm_INNOPATHS_adj_coeff_multiplier | renamed to   cm_adj_coeff_multiplier |   |  
cm_INNOPATHS_inco0Factor | renamed to   cm_inco0Factor |   |  
cm_INNOPATHS_CCS_markup | renamed to   cm_CCS_markup |   |  
cm_INNOPATHS_Industry_CCS_markup | renamed to   cm_Industry_CCS_markup |   |  
cm_INNOPATHS_renewables_floor_cost | renamed to   cm_renewables_floor_cost |   |  
cm_INNOPATHS_DAC_eff | renamed to   cm_DAC_eff |   |  
cm_INNOPATHS_sehe_upper | renamed to   cm_sehe_upper |   |  
cm_INNOPATHS_priceSensiBuild | renamed to   cm_priceSensiBuild |   |  
cm_INNOPATHS_pushCalib | renamed to   cm_pushCalib |   |  
cm_INNOPATHS_reducCostB | renamed to   cm_reducCostB |   |  
cm_INNOPATHS_effHP | renamed to   cm_effHP |   |

### CODE REVIEW

**Comment by Renato-Rodrigues** [comment]
1. Improved regional emission tax convergence algorithm:

* the new algorithm uses the previous iteration model abatement information to estimate the necessary tax update for the next iteration in order to achieve the desired emission target.
   * Big improvement in convergence iterations: convergence of single regional targets can be achieved with around four iterations instead of previously 20+ required iterations.
 * Improved consistence of multiple targets set for the same region (e. 2030, 2040 and 2050 emission targets).
   * a sequential target solution logic solution was implemented to better guide the model through multiple period targets.
     *  Example:
       * Targets set for 2030, 2040 and 2050 for the same region are initially solved by the algorithm sequentially. 
       * Tax prices are re-scaled to minimize the 2030 target deviations. The algorithm assumes a predefined behavior for prices after 2030 based on the projection of previous year prices. This behavior is applied until the next defined target year (2040 in the example case). After the next target year the tax is increased according the cm_postTargetIncrease value.
       * Only after the 2030 target is achieved the algorithm moves to solve the next target, and so on (2040 and 2050 in this case). 
       * After all yearly targets converge sequentially, the algorithm minimizes all target deviations together.
       * At this later stage, priority can still be given to short term targets in order to minimize price trajectory fluctuations. 
       * You can set the switch `cm_prioRescaleFactor` to any number between 0 and 1. Default value = `off`. Recommended value for multiple targets applied to the same region: `0.2`. This value prioritize 2030 target adjustments by a factor of five in comparison with 2050 target adjustments.   
  * Additional option: added option to delay price response for later year targets. 
    * Example:
      * You can set a delay in years to latter targets, keeping the previous price update tendency longer than the target year. 
      * ex: if you set an emission target to be active between 2030 and 2050, with a `cm_emiMktTargetDelay` value of 10 years, the tax updated trajectory is only rescaled for the years greater or equal to 2040 (2030+cm_emiMktTargetDelay). This simulate a delayed price response for later targets, the 2050 set target in this case.
      * default value is equal to `0` years of delayed action. A value of zero means that the 2050 target deviations will adjust the tax trajectory for the entire period (2030 to 2050 in the example above).
      * This option has no effect to the previous target set for an specific region, as a tax trajectory tendency is still defined.

 * code practical changes:
   * `cm_emiMktTarget` replaces `cm_regiCO2target`.
      * It now includes an additional dimension to determine if the target is a economy wide target - `all` - or a emission market specific target - `ETS`, `ESR` or `other`.
   * `cm_postTargetIncrease` default value changed to 0 (from previous 2€ per year value).
   * new switches added: `cm_emiMktTargetDelay` and  `cm_prioRescaleFactor`.
   * deprecated switches: `cm_regiCO2target`, `cm_emiMktETS`, `cm_emiMktETS_type`, `cm_ETS_postTargetIncrease`, `cm_ETS_post2055Increase`, `cm_emiMktES`, `cm_emiMktES_type`, `cm_ESD_postTargetIncrease`, `cm_ESD_post2055Increase`, `cm_emiMktEScoop`, `cm_emiMktES2020price`, `cm_emiMktES2050`.
   * `cm_regipol_slope_beforeTarget` code is obsolete and it was removed from the regipol module.

**Comment by Renato-Rodrigues** [comment]
2. PE, SE, FE and CCS targets achievable by policy taxes or subsidies.: 

* Possibility to check how much the model needs of taxes or subsidies to achieve specific targets.
* The target can be defined at different:
   * quantity targets: primary energy, secondary energy and final energy use, including or not bunkers and non energy use; or CCS
   * groups: biomass, fossil, VRE, renewables, renewablesNoBio, synthetic, hydrogen, electricity, heat technologies or all. (in the future it can be extended to work also at energy carrier level).

 * code practical changes:
   * new switches: `cm_implicitQttyTarget`, `cm_loadFromGDX_implicitQttyTargetTax`
   * deprecated switches: `cm_implicitFE`, `cm_implFETarget`, `cm_implFEExoTax`

  * new switch: `cm_implicitQttyTarget` default = off. Define quantity target for primary, secondary, final energy or CCS (PE, SE and FE in TWa, or CCS in Mt CO2) per target group (total, biomass, fossil, VRE, renewables, synthetic, ...). The target is achieved by an endogenous calculated markup in the form or a tax or subsidy in between iterations. 
     * Ex: If you want to enforce a tax (tax) that guarantees that the total (t=total) Final Energy (FE.all) in 2030 (2030) is at most the Final energy target in the Fit For 55 regulation in the European Union (EU27_regi) (1.03263 Twa). Just set the cm_implicitQttyTarget to "2030.EU27_regi.tax.t.FE.all 1.03263". The p47_implicitQttyTargetTax parameter will contain the tax necessary to achieve that goal. (777.8 Mtoe = 777.8 * 1e6 toe = 777.8 * 1e6 * 41.868 GJ = 777.8 * 1e6 * 41.868 * 1e-9 EJ = 777.8 * 1e6 * 41.868 * 1e-9 * 0.03171 TWa = 1.03263 TWa)   
     * Ex 2: If you want to enforce a subsidy (sub) that guarantees a minimum share (s) of electricity in final energy (FE.electricity) equal to 80% (0.8) from 2050 (2050) onward in all World (GLO) regions. Just set the cm_implicitQttyTarget to "2050.GLO.sub.s.FE.electricity 0.8". The p47_implicitQttyTargetTax parameter will contain the subsidy necessary to achieve that goal.    
  * new switch: `cm_loadFromGDX_implicitQttyTargetTax` -> load p47_implicitQttyTargetTax values from gdx for first iteration on policy scenarios.

**Comment by Renato-Rodrigues** [comment]
3. PE and FE exogenous defined prices: 

* Added support for PE or FE exogenously defined prices. 
* It can be used to test the model behavior if a certain energy carrier has an specific price. The algorithm calculates and update endogenously an "implicit" tax or subsidy so the model always sees the desired exogenous price at its optimization decision.
* Check the files `.\modules\47_regipol\regiCarbonPrice\input\exogenousFEprices.cs3r` or `.\modules\47_regipol\regiCarbonPrice\input\exogenousPEprices.cs3r` to see examples on how to define an exogenous price scenario to be loaded into the parameter `pm_implicitPriceTarget` to use this feature. 

 * code practical changes:
   * new switches: `cm_implicitPriceTarget`, `cm_implicitPePriceTarget`

**Comment by Renato-Rodrigues** [comment]
4. Added support for defining minimum VRE shares or maximum CCS values in scenarios

* these switches only work on native remind regions, i.e. if you want to set a region group bound, like for example an EU27 target in a REMIND 21 regions run, please use the `cm_implicitQttyTarget` switch instead.

* new switch: `cm_VREminShare`. 
   * Default=off.
   * E.g. `2050.EUR_regi 0.7`, requires a minimum 70% VRE share (wind plus solar) in electricity production.

* new switch: `cm_CCSmaxBound`. 
   * Default=off.
   * E.g. `GLO 2, EUR 0.25`, Carbon Capture and Storage (including DACCS and BECCS) is limited to a maximum of 2GtCO2/yr globally, and 250 Mt CO2/yr in EU28.

**Comment by Renato-Rodrigues** [comment]
5. Marginal prices after taxes

* Added additional FE balance to allow saving marginal costs including FE, SE, sector and emission market information.
* These prices are reported as marginal prices in the remind2 library.

**Comment by Renato-Rodrigues** [comment]
6. Code structure improvements:

* removing deprecated scenario config files. 
   * REMIND_EU_calibration and sensitivity scenario configs were removed. 
     * You should use instead the scenario config Fit_for_55_sensitivity if you were using those files;
   * INNOPATHS scenario config file was removed. 
* removing deprecated ETS and ESR switches from existent scenario config files.
* updating existent scenario config files switches to be compatible with latest changes (ex: replacing `cm_regiCO2target` by the new `cm_emiMktTarget` definition).
* removing deprecated implicit FE switches and code.
  * You should use the improved formulation contained in the `cm_implicitQttyTarget` switch instead.  
* rewriting entirely the module 47 code oriented to more performance and more structured code necessary to better handle multiple regional targets and future requirements from the ARIADNE and ECEMF projects. 
* replaced equations in the module 47 by parameters calculated in the postsolve to improve the model memory footprint.
* removed project specific references from general switch names (e.g. removing references to the INNOPATHS project from the model and scenario config files).

**Comment by Renato-Rodrigues** [comment]
@robertpietzcker @fschreyer @lavinia I consider this code ready to be merged with remind. 
I am running during this weekend final calibration and test runs with the default configs for H12 and EU21 before I merge it with the trunk.
I would highly appreciate if you could find some time to take a look at this PR and bring any points and or objections with it before this is merged to remind. 
As you can see, there is a quite extensive list of changes done to the code at once. Unfortunately, I would waste too much time splitting each one of those changes into separate PRs.
I am fully available to explain or discuss anything necessary in detail.
and thank you in advance for the review!

**Comment by Renato-Rodrigues** [comment]
All tests are finished and successful: `/p/projects/remind/users/renatoro/ECEMF/v07_2022_10_07_H12/output`
I am just waiting for your approval to merge these code changes with remind.

**Review by LaviniaBaumstark — COMMENTED** [review]
* one general comment: could we get rid of teh EU-specific subfolder in the config folder an have all needed scenario_config and mapping files directly in the config folder?
* why do you need s many new interfaces?
* is transport complex working now and generating meaningful results?

**Comment by Renato-Rodrigues** [comment]
> * one general comment: could we get rid of teh EU-specific subfolder in the config folder an have all needed scenario_config and mapping files directly in the config folder?

Yes, I can move the config files, but I would prefer to make it in a separate PR after this one is merged.

> * why do you need s many new interfaces?

I am not sure what do you refer by interfaces. If you are referring to the pm_ and vm_ parameters and variables, I followed the code check requirements for them. 
There are quite a few of those interfaces between the regipol and the optimization module because any in-between-iteration algorithm need to have a convergence check in the optimization module. This is necessary to inform the model if you need additional iterations to assure that the regipol specific features also converged.
As the display of reasons for convergence happens in the optimization model, there are quite a few o debugging displays there that requires a parameter from another module. 

> * is transport complex working now and generating meaningful results?

Yes, transport complex is working fine with the current input data revision (6.316) and with the upcoming one (6.322). 
When the input data revision is updated to 6.322, alois PR (https://github.com/remindmodel/remind/pull/996) needs to be merged to keep everything working fine.
I am also waiting the input data revision update and clarifying the calibration script status ("collect_calibration") to update the calibration files in the cluster, but you can find the calibration in the folder I posted above for EU21.

File: core/bounds.gms
Comment (Renato-Rodrigues) [reply core/bounds.gms:234]: yes. This relax the bound to work better with smaller regions allowing some flexibility for ESW and ECE regions in EU21 regions disaggregation.

File: modules/29_CES_parameters/calibrate/datainput.gms
Comment (Renato-Rodrigues) [reply modules/29_CES_parameters/calibrate/datainput.gms:259]: The switch is used for alternative industry FE trajectories in ARIADNE ETS runs and ECEMF runs.
This code could be removed from the remind code as soon as:
 (1) industry subsectors FE trajectories are translated into industry fixed shares dimensions automatically in mrremind code and 
  (2) we validate industry trajectories at EU country level.

Comment (LaviniaBaumstark) [reply modules/29_CES_parameters/calibrate/datainput.gms:259]: o.k.

**Review by LaviniaBaumstark — APPROVED** [review]

**Review by fschreyer — APPROVED** [review]
Ok, seems a lot and I had a rough glance. I would actually be interested in the results of these new FE prices we have. For example, whether ``p_FEPrice_by_SE_Sector_EmiMkt`` is actually different for different SEs, where the SEs are perfect substitutes. We can look at this some time.

File: modules/31_fossil/timeDepGrades/equations.gms
Comment (Renato-Rodrigues) [reply modules/31_fossil/timeDepGrades/equations.gms:179]: yes

**Pull request merged**
